### DISCUSSION TITLE
Dynamic assetPrefix for ISR pages (custom server)

### DISCUSSION THREAD

**Discussion by flipflopsandrice** [vercel/next.js#18301] [Help] [answered]
# Bug report

## Describe the bug

Pages that are marked as ISR are built with a static `assetPrefix` upon first load.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Create an ISR page (`getStaticPaths` with `fallback: true`, and `getStaticProps` with `revalidate: x`)
2. Create a custom server that sets a dynamic assetPrefix (``app.setAssetPrefix(`/assets/${process.env.REGION}`);``)
3. Build/run server with `export REGION=eu .....`
4. Load page and see static assets loaded from `/_next/static/**/*`)
5. Reload page and see static assets loaded from `/assets/eu/_next/static/**/*`

## Expected behavior

We expect the ISR pages to have the correct assetPrefix upon first load, since we are waiting for `getStaticProps` anyway, thus no need for a build-time assetPrefix.

## System information

- Version of Next.js: 9.5.5

## Additional context

Our use-case is we prepend our assets with a specific region tag so we can run different releases in different CDN clusters (eg: `/nl/` page serves assets through `/assets/eu/_next/static/**/*`).

**Comment by flipflopsandrice** [comment]
For now we have created a pre-start script that patches the built artifacts with the fixed `assetPrefix`:

From: `/_next/static/...`
To: `/assets/eu/_next/static/...`

*next.config.js*:
```
module.exports = {
    assetPrefix: 'BUILDTIME_ASSETPREFIX',
    ....
};
```

*`yarn start` also runs:*
```
#!/usr/bin/env bash

if [ -z "$REGION" ]; then
    # Bail out with a warning in case the variable is missing
    echo "WARNING: \"REGION\" environment variable not present"
    exit 0;
fi

script_full_path=$(dirname "${BASH_SOURCE[0]}")
script_full_path=$(cd "$script_full_path/.." && pwd)
find "$script_full_path/".next/ -type f -print0 | \
    xargs -0 -n 1 sed -i 's#\"BUILDTIME_ASSETPREFIX#\"\/assets/'"${REGION,,}"'#g' $1

```

**Comment by flipflopsandrice** [comment, marked as answer]
It seems NextJS 10's `fallback: "blocking"` might actually fix this :+1:, since it doesn't send a static fallback to browser.

**Reply by fbudassi** [reply]
Do you know how "unstable" is the "unstable_blocking" in Next.js 9.5?
For now, I'm unable to upgrade to next.js 10.

**Reply by daomul** [reply]
"Revalidate" is also used, but after configuring assetprefix CDN, the static resources in JS / CSS and "public" folder do not follow the CDN prefix. Do you have the same problem?

**Comment by VojtechVitek** [comment]
I like what you did there @flipflopsandrice. I've adapted this to our Dockerfile:

next.config.js:
```js
const config = {
   // ...
   assetPrefix: process.env.NODE_ENV === 'production' ? "NEXT_ASSET_PREFIX_CDN_URL_REPLACE_ME" : '',
   // ...
}
```

Dockerfile:
```Dockerfile
# Runtime CDN URL to serve /_next/* static assets.
ENV CDN_URL https://cdn.example.com

CMD sh -c 'find build/ -type f -print0 | xargs -0 -n 1 sed -i s#NEXT_ASSET_PREFIX_CDN_URL_REPLACE_ME#$CDN_URL#g $1; exec node --max-old-space-size=4096 --max-http-header-size=80000 ./build/index.js'
```

**Comment by ScriptedAlchemy** [comment]
```js
if you use a custom server, after calling next(), you can reset the asset path at any time 
  if (assetPrefix) {
    nextApp.setAssetPrefix(assetPrefix);
  }
```

**Reply by justrhysism** [reply]
Ohhh this was a real nugget of gold 🌟 which has enabled us to move toward immutable builds. Underrated. Thanks @ScriptedAlchemy!

**Reply by CMCDragonkai** [reply]
This no longer works for NextJS 13 right? I just tried it, and the `href` paths don't change.

**Comment by daomul** [comment]
您好，您的邮件已收到，尽快给您回复。

**Comment by LukasBombach** [comment]
Has anyone attempted to overwrite the `script`-related outputputs in Next's `<Head />` and `<NextScript />`?

**Comment by daomul** [comment]
您好，您的邮件已收到，尽快给您回复。

**Reply by LukasBombach** [reply]
^ This translates to 

>Hello, your email has been received, I will reply to you as soon as possible.

**Comment by daomul** [comment]
您好，您的邮件已收到，尽快给您回复。

**Comment by daomul** [comment]
您好，您的邮件已收到，尽快给您回复。
